Plugins/DuplicationPlugin/DuplicateContextMenuItem.cs

```diff
@@ -1,156 +1,651 @@
-                        key = "null";
-                    extensions[key].DuplicateAsset(entry, newName, newType != null, newType);
-                }
-                catch (Exception e)
-                {
-                    App.Logger.Log($"Failed to duplicate {entry.Name}");
-                }
-            });
-
-            App.EditorWindow.DataExplorer.RefreshAll();
-        });
-    }
-}
+﻿using AtlasTexturePlugin;
+using DuplicationPlugin.Windows;
+using Frosty.Core;
+using Frosty.Core.Viewport;
+using Frosty.Core.Windows;
+using Frosty.Hash;
+using FrostySdk;
+using FrostySdk.Ebx;
+using FrostySdk.IO;
+using FrostySdk.Managers;
+using FrostySdk.Resources;
+using MeshSetPlugin.Resources;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Security.Cryptography;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace DuplicationPlugin
+{
+    public class DuplicationTool
+    {
+        #region --Extensions--
+
+        public class SoundWaveExtension : DuplicateAssetExtension
+        {
+            public override string AssetType => "SoundWaveAsset";
+
+            public override EbxAssetEntry DuplicateAsset(EbxAssetEntry entry, string newName, bool createNew, Type newType)
+            {
+                EbxAssetEntry refEntry = base.DuplicateAsset(entry, newName, createNew, newType);
+
+                EbxAsset refAsset = App.AssetManager.GetEbx(refEntry);
+                dynamic refRoot = refAsset.RootObject;
+
+                foreach (dynamic chkref in refRoot.Chunks)
+                {
+                    ChunkAssetEntry soundChunk = App.AssetManager.GetChunkEntry(chkref.ChunkId);
+                    ChunkAssetEntry newSoundChunk = DuplicateChunk(soundChunk);
+                    chkref.ChunkId = newSoundChunk.Id;
+                }
+                App.AssetManager.ModifyEbx(refEntry.Name, refAsset);
+
+                return refEntry;
+            }
+        }
+
+        public class PathfindingExtension : DuplicateAssetExtension
+        {
+
+            public override string AssetType => "PathfindingBlobAsset";
+
+            public override EbxAssetEntry DuplicateAsset(EbxAssetEntry entry, string newName, bool createNew, Type newType)
+            {
+                EbxAssetEntry refEntry = base.DuplicateAsset(entry, newName, createNew, newType);
+
+                EbxAsset refAsset = App.AssetManager.GetEbx(refEntry);
+                dynamic refRoot = refAsset.RootObject;
+                ChunkAssetEntry pathfindingChunk = App.AssetManager.GetChunkEntry(refRoot.Blob.BlobId);
+                if (pathfindingChunk != null)
+                {
+                    ChunkAssetEntry newPathfindingChunk = DuplicateChunk(pathfindingChunk);
+                    if (newPathfindingChunk != null)
+                    {
+                        refRoot.Blob.BlobId = new Guid(newPathfindingChunk.Name);

```

Comment:
same thing here use chunk.Id

Reply:
but yeah that should be it, couldnt find anything else which needs to be changed, if u change this let me know so i can merge it